### ISSUE TITLE
Make Arroyo opinionated in resource lifecycle management

### ISSUE THREAD

**Issue by fpacifici** [getsentry/arroyo#126]
There is no clear direction suggested by the API on how to manage the lifecycle of stateful resources like a `Producer`.
This introduced a variety of different approaches in production that are not consistent with each other.
Most of them are broken in a way or another (for example we have consumers that fail to close producers).
This is made worse as getting the lifecycle right is hard since most people do not properly grasp the lifecycle of a ProcessingStrategy to begin with.

There are fundamentally three options:
- If a ProcessingStrategy needs a Producer it can instantiate it, own it, and close it. The processing strategy is fully responsible.
- If a ProcessingStrategy needs a Producer it can rely on the callsite that instantiate the factory to provide a functioning Producer. The callsite is fully responsible and the processing strategy leaves it open.
- If a ProcessingStrategy needs a Producer it can rely on the callsite to provide a Producer and it gains ownership at that point, thus close it at the end.

I think the third will be massively confusing (also fairly hard to achieve consistently as the ProcessingStrategy can be closed at every rebalance).
The first is problematic for unit tests as of today. If the processing strategy instantiates a Producer it is impossible to use the Fake Producer on unit tests. This cannot be alleviated by creating the producer in the Factory as the factory cannot close it.

Without fundamental changes in Producer instantiation, we would have to rely on option two. 

So this issue can be addressed in two ways:
- [enforce Options 2]: Investigate ways to detect when a client tries to use option 1 or 3 and block it. Or restructure the Producer creation process to make it harder to implement 1 or 3. I think the way to address this is, first, to see if we can infer the cases where 1 or 3 are being used automatically.
- [fundamental changes in Producer instantiation]: Can we study ways to make the API responsible to provide the needed resources given a spec from the client? So the API would have the control of the Producer lifecycle in a way that can be customized for unit tests. Example: provide a spec for the producer needed by a strategy factory. The processor keeps a registry of resources. The strategy factory retrieves the producer instance when it needs it via a context manager. The processor decides when it is time to close it.

**Comment by lynnagara** [comment]
I'm not really convinced this is a real problem or we should be trying to force consistency here.

The problem of who is responsible for the lifecycle of resources applies to all resources, of which KafkaProducer is just one. Of course the difference is that KafkaProducer happens to be provided by this library, but the same fundamental issue
applies to any resources that need opening/closing - threadpools, database connections, or whatever custom resource your code uses. I think it would be very unexpected for KafkaProducer to not adhere to the same way of resource management as everything else, and we should not try and treat it in any special way. 

As for the 3 options outlined above: Option 3 is wrong. In the strategies provided by Arroyo out of the box, Arroyo will never do option 3. It will either do option 1 - fully manage the resources (e.g. threadpool created and closed by RunTaskInThreads strategy), or option 2 - it gets passed a resource (e.g. ProduceStep gets passed a producer) and will not close it in that scenario since it didn't create it.

Now, there are reasons why it is done differently in these 2 cases. We want the user to be able to easily pass us a custom producer with whatever settings they want on it, and we also want to be able to swap it out for the LocalProducer to allow unit testing without Kafka. I don't really think this difference is a problem, nor that we should force every resource to be managed in the same way since we know that these 2 scenarios exist.

To me, there is a simple rule: if you created a resource you close it, and if Arroyo created it it closes it for you. I think it's fairly consistent. As far as I know Arroyo already follows this pattern today.

One tangential comment on the specific case of closing producers. Arroyo's ProduceStep always makes sure the futures have all completed before the strategy closes. Though people should of course be closing their resources, not doing so in this particular case is fairly harmless since Arroyo ensures there won't be pending messages that need to get flushed.

**Comment by fpacifici** [comment]
> To me, there is a simple rule: if you created a resource you close it, and if Arroyo created it it closes it for you. I think it's fairly consistent. As far as I know Arroyo already follows this pattern today.

I agree with that, though now we have observed that this concept seem to be fairly error prone. I think I remember lifecycle related comments appearing in pretty much every PR that was adding a consumer. This, to me, should be a hint that we can improve the api to make it harder to do the wrong thing.

Let's say that the direction provided by the library is "the Producer is a resource and you have to close it accordingly". Now there is something to improve:
I don't think people (especially those who are new to the api)  have a good grasp about the lifecycle of a processing strategy. Specifically I see people having a hard time in connecting the `join` method with the destruction of the strategy. This causes resources not to be closed. Also I see people having a hard time getting that a strategy can be rebuilt (rebalance) without the consumer being terminated, thus closing a resource that was instantiated only once too early (which basically is approach three above).

We could start by improving the documentation on the life cycle and push people to use the ProduceAndCommit step.
